prov/shm/src/smr_dsa.c

```diff
@@ -0,0 +1,848 @@
+/*
+ * Copyright (c) 2022 Intel Corporation. All rights reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include "smr.h"
+
+#if SHM_HAVE_DSA
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdatomic.h>
+#include <sys/un.h>
+#include <accel-config/libaccel_config.h>
+#include <linux/idxd.h> // Used for dsa_hw_desc and dsa_completion_record structs
+#include "ofi_shm.h"
+#include "smr_dsa.h"
+
+#define MAX_WQS_PER_EP 4
+#define MAX_BITMAP_SIZE 32
+#define GENCAP_CACHE_CTRL_MEM 0x4
+#define LIMITED_MSIX_PORTAL_OFFSET 0x1000
+
+#define CMD_CONTEXT_COUNT MAX_BITMAP_SIZE
+
+// Numbe of DSA descriptors must be large enough to fill the
+// maximum number of SAR buffers.
+#define MAX_CMD_BATCH_SIZE (SMR_BUF_BATCH_MAX + SMR_IOV_LIMIT)
+
+struct dsa_bitmap {
+	int size;
+	atomic_int data;
+};
+
+enum {
+	DSA_CMD_COMPLETED,
+	DSA_CMD_NOT_COMPLETED,
```

Comment:
Do you need an enum for a bool that you use as a return in one function?

Reply:
We will change it to use a bool instead